File: javascript/queues/queue.js
class Queue {
  constructor() {
    this.items = {};
    this.frontIndex = 0;
    this.rearIndex = 0;
    // this.size = 0;
  }
  enqueue(item) {
    this.items[this.rearIndex++] = item;
  }
  dequeue() {
    if (this.is_empty()) {
      throw new Error("Queue is empty");
    }
    const item = this.items[this.frontIndex];
    delete this.items[this.frontIndex];
    this.frontIndex++;
    console.log("DEQUEUE: ", item);

    return item;
  }
  peek() {
    if (this.is_empty()) {
      return "Queue is emty";
    }
    return this.items[this.frontIndex];
  }
  is_empty() {
    return this.rearIndex - this.frontIndex === 0;
  }
  size() {
    return this.rearIndex - this.frontIndex;
  }
}
// module.exports = Queue;

const queue = new Queue();

queue.enqueue(10);
queue.enqueue(20);
queue.enqueue(30);
// queue.enqueue(40);
// console.log(queue);
// queue.dequeue();
// queue.dequeue();
// queue.dequeue();
console.log(queue);

File: python/queues/queue.py
class ArrayQueue:
    def __init__(self):
        self.items={}
        self.front_index=0
        self.rear_index=0
    def enqueue(self,item):
        self.items[self.rear_index] = item
        self.rear_index += 1
    def dequeue(self):
        if self.is_empty():
            raise Exception('Queue is empty')
        item = self.items[self.front_index]
        del self.items[self.front_index]
        self.front_index += 1
        return item
    def peek(self):
        if self.is_empty():
            raise Exception('Queue is empty')
        return self.items[self.front_index]
    def is_empty(self):
        return self.rear_index - self.front_index == 0
    def size(self):
        return self.rear_index - self.front_index
    def __str__(self):
        return str([self.items[i] for i in range(self.front_index, self.rear_index)])

q = ArrayQueue()
q.enqueue(10)
q.enqueue(20)
q.enqueue(30)

print(q)          # [10, 20, 30]
print(q.dequeue())  # 10
print(q)          # [20, 30]
print(q.peek())   # 20
